# In-sample fitting and out-of-sample predictive performance of Random Forests, Neural Networks and Support Vector Machines by using two different cross-validation techniques. 

Extracts data and saves it in loan

In [117]:
loan=read.csv("c:/Users/bibek/Downloads/loan.csv")

Displays the number of fields and records in the dataset

In [118]:
dim(loan)

[1] 850   6

In [119]:
names(loan)


[1] "Status"       "CreditGrade"  "Amount"       "Age"          "BorrowerRate"
[6] "DebtToIncome"

Encodes the vector as a factor

In [120]:
is.factor(loan$CreditGrade)

[1] TRUE

Randomizes the data set and sets it to 1

In [121]:
set.seed(1)

Training set is defined with 750 datasets

In [122]:
train=sample(1:850,750)

Installs and loads the random forest library

In [123]:
library(randomForest)

In [124]:
rf=randomForest(factor(Status) ~ ., data=loan)

Performance of random forest in 850 dataset

In [125]:
rf


Call:
 randomForest(formula = factor(Status) ~ ., data = loan) 
               Type of random forest: classification
                     Number of trees: 500
No. of variables tried at each split: 2

        OOB estimate of  error rate: 10.24%
Confusion matrix:
    0   1 class.error
0 392  33  0.07764706
1  54 371  0.12705882

In [126]:
rf=randomForest(factor(Status) ~ ., data=loan[train,])


Performance of random forest in 750 train dataset

In [127]:
 rf


Call:
 randomForest(formula = factor(Status) ~ ., data = loan[train,      ]) 
               Type of random forest: classification
                     Number of trees: 500
No. of variables tried at each split: 2

        OOB estimate of  error rate: 10.67%
Confusion matrix:
    0   1 class.error
0 343  35  0.09259259
1  45 327  0.12096774

In [128]:
pre.rf=predict(rf,loan[-train,])


Performance of random forest in 100 test dataset

In [129]:
table(loan$Status[-train],pre.rf)

   pre.rf
     0  1
  0 44  3
  1  6 47

# In sample Prediction 

Fits the dataset into Neural Network with six neurons and maxit iteration of 1000.

In [130]:
library(nnet)


In [131]:
X=model.matrix(Status~.,data=loan)[,-1]

In [132]:
XX=cbind(Status=loan$Status,scale(X))

In [133]:
loannet= nnet(factor(Status)~., data=XX,  size = 6,maxit=1000)


# weights:  49
initial  value 696.447373 
iter  10 value 333.980200
iter  20 value 284.559503
iter  30 value 263.256956
iter  40 value 242.875629
iter  50 value 236.388429
iter  60 value 230.816745
iter  70 value 228.820787
iter  80 value 227.361329
iter  90 value 227.000942
iter 100 value 226.523450
iter 110 value 226.338091
iter 120 value 226.116947
iter 130 value 225.934701
iter 140 value 223.768175
iter 150 value 222.064257
iter 160 value 220.956505
iter 170 value 220.358005
iter 180 value 217.031549
iter 190 value 213.728101
iter 200 value 212.559190
iter 210 value 212.495464
iter 220 value 212.387788
iter 230 value 212.290795
iter 240 value 212.209444
iter 250 value 211.880554
iter 260 value 211.607459
iter 270 value 211.523748
iter 280 value 211.433920
iter 290 value 211.373422
iter 300 value 211.333558
iter 310 value 211.327979
iter 320 value 211.290079
iter 330 value 211.229982
iter 340 value 211.216031
iter 350 value 211.209457
iter 360 value 211.181389
iter 370 value 211.154

In [134]:
pred=predict(loannet,XX,type="class")

In [135]:
table(loan$Status,pred)

   pred
      0   1
  0 395  30
  1  50 375

770 out of 850 dataset are correctly predicted

# Out of sample prediction 

In [136]:
loannet2= nnet(factor(Status)~., data=XX[train,],  size = 6,maxit=1000)

# weights:  49
initial  value 527.463768 
iter  10 value 244.496124
iter  20 value 208.583995
iter  30 value 183.978937
iter  40 value 174.542452
iter  50 value 172.126882
iter  60 value 170.148798
iter  70 value 168.576759
iter  80 value 167.240466
iter  90 value 162.653874
iter 100 value 160.312421
iter 110 value 158.611813
iter 120 value 154.377751
iter 130 value 151.659140
iter 140 value 151.334552
iter 150 value 151.042370
iter 160 value 150.960327
iter 170 value 150.938423
iter 180 value 150.924292
iter 190 value 150.913064
iter 200 value 150.902671
iter 210 value 150.897224
iter 220 value 150.890930
iter 230 value 150.889510
iter 240 value 150.888366
iter 250 value 150.886995
final  value 150.885925 
converged


In [137]:
pred=predict(loannet2,XX[-train,],type="class")

In [138]:
table(loan$Status[-train],pred)


   pred
     0  1
  0 41  6
  1 10 43

84 out of 100 training set are correctly classified  by ANN

Uses 750 randomly selected test dataset to build a model in Support Vector Machine

In [139]:
library(e1071)

In [140]:
sv=svm(factor(Status) ~ ., data = loan[train,])

Applies the built SVM model in training set

In [141]:
pred = predict(sv, loan[-train,])

In [142]:
table(pred, loan$Status[-train])

    
pred  0  1
   0 41 10
   1  6 43

84 out of 100 training set are correctly classified by SVM

Randomizes the order of rows

In [143]:
set.seed(2)

In [144]:
loan=loan[sample(850),]

# 10 fold cross-validation using Random Forest

In [145]:
apre.rf=0

In [146]:
for(k in 1:10) {
hout=(85*(k-1)+1):(85*k)
loan.rf=randomForest(factor(Status) ~ ., data=loan[-hout,])
pre=predict(loan.rf,loan[hout,])
apre.rf=apre.rf+sum(loan$Status[hout]==pre)
}

In [147]:
apre.rf


[1] 761

761 out of 850 total data set are correctly classified by Random Forest

# 10 fold cross-validation using ANN

In [148]:
apre.nn=0


In [149]:
X=model.matrix(Status~.,data=loan)[,-1]

In [150]:
XX=cbind(Status=loan$Status,scale(X))

In [151]:
for(k in 1:10) {
hout=(85*(k-1)+1):(85*k)
loan.nn=nnet(factor(Status) ~ ., data=XX[-hout,], size=6, maxit=1000)
pre=predict(loan.nn,XX[hout,],type="class")
apre.nn=apre.nn+sum(loan$Status[hout]==pre)
}

# weights:  49
initial  value 538.302995 
iter  10 value 252.965656
iter  20 value 210.714023
iter  30 value 182.924391
iter  40 value 178.701223
iter  50 value 175.857287
iter  60 value 172.666147
iter  70 value 168.897621
iter  80 value 166.800057
iter  90 value 166.180713
iter 100 value 164.652946
iter 110 value 164.410187
iter 120 value 164.038612
iter 130 value 163.790406
iter 140 value 163.691436
iter 150 value 163.581399
iter 160 value 163.338403
iter 170 value 162.823819
iter 180 value 162.040774
iter 190 value 161.078442
iter 200 value 160.821820
iter 210 value 160.622649
iter 220 value 160.489425
iter 230 value 160.010834
iter 240 value 159.233223
iter 250 value 158.755076
iter 260 value 158.593109
iter 270 value 158.416356
iter 280 value 158.292654
iter 290 value 158.230907
iter 300 value 158.218287
iter 310 value 158.213594
iter 320 value 158.210009
iter 330 value 158.183728
iter 340 value 158.156177
iter 350 value 158.154322
iter 360 value 158.153198
iter 370 value 158.149

initial  value 560.083276 
iter  10 value 287.035284
iter  20 value 253.227618
iter  30 value 219.063752
iter  40 value 206.483398
iter  50 value 182.891330
iter  60 value 176.666400
iter  70 value 172.422579
iter  80 value 166.578935
iter  90 value 163.565172
iter 100 value 160.473083
iter 110 value 159.210617
iter 120 value 158.058730
iter 130 value 155.925692
iter 140 value 154.638358
iter 150 value 153.680623
iter 160 value 151.902703
iter 170 value 150.084014
iter 180 value 148.294618
iter 190 value 147.193654
iter 200 value 146.318285
iter 210 value 146.244493
iter 220 value 146.210600
iter 230 value 146.190995
iter 240 value 146.171104
iter 250 value 146.127153
iter 260 value 146.113574
iter 270 value 146.108506
iter 280 value 146.098065
iter 290 value 146.093659
iter 300 value 146.088600
iter 310 value 146.066603
iter 320 value 145.996380
iter 330 value 145.946178
iter 340 value 145.918864
iter 350 value 145.907927
iter 360 value 145.903396
iter 370 value 145.899903
iter 380 va

In [152]:
apre.nn


[1] 718

718 out of 850 total data set are correctly classified by ANN

# 10 fold cross-validation using SVM

In [153]:
apre.sv=0

In [154]:
for(k in 1:10) {
hout=(85*(k-1)+1):(85*k)
loan.sv=svm(factor(Status) ~ ., data=loan[-hout,])
pre=predict(loan.sv,loan[hout,])
apre.sv=apre.sv+sum(loan$Status[hout]==pre)
}

In [155]:
apre.sv


[1] 727

727 out of 850 total data set are correctly classified by SVM.

# Leave one out cross-validation using RandomForest

In [156]:
apre=0

In [157]:
for(k in 1:850) {
loan.rf=randomForest(factor(Status) ~ ., data=loan[-k,])
pre=predict(loan.rf,loan[k,])
if (loan$Status[k]==pre){
apre=apre+1}
}


In [158]:
apre



[1] 762

762 out of 850 total data set are correctly classified based on Leave one out cross-validation.